criteria.c

```diff
@@ -0,0 +1,317 @@
+#define _POSIX_C_SOURCE 200809L
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include <strings.h>
+#include <wayland-client.h>
+#include "mako.h"
+#include "config.h"
+#include "criteria.h"
+#include "notification.h"
+
+struct mako_criteria *create_criteria(struct mako_config *config) {
+	struct mako_criteria *criteria = calloc(1, sizeof(struct mako_criteria));
+	if (criteria == NULL) {
+		fprintf(stderr, "allocation failed\n");
+		return NULL;
+	}
+
+	wl_list_insert(config->criteria.prev, &criteria->link);
+	return criteria;
+}
+
+void destroy_criteria(struct mako_criteria *criteria) {
+	wl_list_remove(&criteria->link);
+
+	finish_style(&criteria->style);
+	free(criteria->app_name);
+	free(criteria->app_icon);
+	free(criteria->category);
+	free(criteria->desktop_entry);
+	free(criteria);
+}
+
+bool match_criteria(struct mako_criteria *criteria,
+		struct mako_notification *notif) {
+	struct mako_criteria_spec spec = criteria->spec;
+
+	if (spec.app_name &&
+			strcmp(criteria->app_name, notif->app_name) != 0) {
+		return false;
+	}
+
+	if (spec.app_icon &&
+			strcmp(criteria->app_icon, notif->app_icon) != 0) {
+		return false;
+	}
+
+	if (spec.actionable &&
+			criteria->actionable == wl_list_empty(&notif->actions)) {
+		return false;
+	}
+
+	if (spec.urgency &&
+			criteria->urgency != notif->urgency) {
+		return false;
+	}
+
+	if (spec.category &&
+			strcmp(criteria->category, notif->category) != 0) {
+		return false;
+	}
+
+	if (spec.desktop_entry &&
+			strcmp(criteria->desktop_entry, notif->desktop_entry) != 0) {
+		return false;
+	}
+
+	return true;
+}
+
+bool parse_boolean(const char *string, bool *out) {
```

Comment:
This could probably be reused in `config.c` too

Reply:
I was planning on moving this and the existing stuff into a `parse.c` utility-type file in another PR, which would include making `config.c` use it. `config.c` is getting a little bit out of control.

Reply:
Yeah, nice idea. So let's keep it that way for now.